src/runtime/cuda.cpp

```diff
@@ -61,46 +61,59 @@ extern "C" WEAK void *halide_cuda_get_symbol(void *user_context, const char *nam
-WEAK void load_libcuda(void *user_context) {
+WEAK int load_libcuda(void *user_context) {
     debug(user_context) << "    load_libcuda (user_context: " << user_context << ")\n";
     halide_abort_if_false(user_context, cuInit == nullptr);
+    halide_error_code_t result;
 
 // clang-format off
-#define CUDA_FN(ret, fn, args)               fn = get_cuda_symbol<ret(CUDAAPI *) args>(user_context, #fn);        // NOLINT(bugprone-macro-parentheses)
-#define CUDA_FN_OPTIONAL(ret, fn, args)      fn = get_cuda_symbol<ret(CUDAAPI *) args>(user_context, #fn, true);  // NOLINT(bugprone-macro-parentheses)
-#define CUDA_FN_3020(ret, fn, fn_3020, args) fn = get_cuda_symbol<ret(CUDAAPI *) args>(user_context, #fn_3020);   // NOLINT(bugprone-macro-parentheses)
-#define CUDA_FN_4000(ret, fn, fn_4000, args) fn = get_cuda_symbol<ret(CUDAAPI *) args>(user_context, #fn_4000);   // NOLINT(bugprone-macro-parentheses)
+#define CUDA_FN(ret, fn, args)               result = get_cuda_symbol<ret(CUDAAPI *) args>(user_context, #fn, false, fn); if (result) return result;        // NOLINT(bugprone-macro-parentheses)
+#define CUDA_FN_OPTIONAL(ret, fn, args)      result = get_cuda_symbol<ret(CUDAAPI *) args>(user_context, #fn, true, fn); if (result) return result; // NOLINT(bugprone-macro-parentheses)
+#define CUDA_FN_3020(ret, fn, fn_3020, args) result = get_cuda_symbol<ret(CUDAAPI *) args>(user_context, #fn_3020, false, fn); if (result) return result;  // NOLINT(bugprone-macro-parentheses)
+#define CUDA_FN_4000(ret, fn, fn_4000, args) result = get_cuda_symbol<ret(CUDAAPI *) args>(user_context, #fn_4000, false, fn); if (result) return result;  // NOLINT(bugprone-macro-parentheses)
 #include "cuda_functions.h"
 #undef CUDA_FN
 #undef CUDA_FN_OPTIONAL
 #undef CUDA_FN_3020
 #undef CUDA_FN_4000
     // clang-format on
+    return halide_error_code_success;
 }
 
 // Call load_libcuda() if CUDA library has not been loaded.
 // This function is thread safe.
 // Note that initialization might fail. The caller can detect such failure by checking whether cuInit is nullptr.
-WEAK void ensure_libcuda_init(void *user_context) {
+WEAK int ensure_libcuda_init(void *user_context) {
     ScopedSpinLock spinlock(&lib_cuda_lock);
     if (!cuInit) {
-        load_libcuda(user_context);
+        return load_libcuda(user_context);
     }
+    return halide_error_code_success;
 }
 
 extern WEAK halide_device_interface_t cuda_device_interface;
 
-WEAK const char *get_error_name(CUresult error);
-WEAK CUresult create_cuda_context(void *user_context, CUcontext *ctx);
+WEAK const char *get_cuda_error_name(CUresult error);
+WEAK int create_cuda_context(void *user_context, CUcontext *ctx);
+
+template<typename... Args>
+int error_cuda(void *user_context, CUresult cuda_error, const Args &...args) {
+    if (cuda_error == CUDA_SUCCESS) {
+        return halide_error_code_success;
+    }
+    error(user_context).append("CUDA error: ", get_cuda_error_name(cuda_error), args...);
```

Comment:
Does this still need a " " between the cuda error name and the rest of the args?

Reply:
Fixed